File: ios/Sources/RynBridgeBackgroundTask/BackgroundTaskModule.swift
import Foundation
import RynBridge

public struct BackgroundTaskModule: BridgeModule, Sendable {
    public let name = "backgroundTask"
    public let version = "0.1.0"
    public let actions: [String: ActionHandler]

    public init(provider: BackgroundTaskProvider) {
        actions = [
            "scheduleTask": { payload in
                guard let taskId = payload["taskId"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: taskId")
                }
                guard let type = payload["type"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: type")
                }
                let interval = payload["interval"]?.intValue
                let delay = payload["delay"]?.intValue
                let requiresNetwork = payload["requiresNetwork"]?.boolValue ?? false
                let requiresCharging = payload["requiresCharging"]?.boolValue ?? false
                let success = try await provider.scheduleTask(
                    taskId: taskId,
                    type: type,
                    interval: interval,
                    delay: delay,
                    requiresNetwork: requiresNetwork,
                    requiresCharging: requiresCharging
                )
                return ["taskId": .string(taskId), "success": .bool(success)]
            },
            "cancelTask": { payload in
                guard let taskId = payload["taskId"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: taskId")
                }
                let success = try await provider.cancelTask(taskId: taskId)
                return ["success": .bool(success)]
            },
            "cancelAllTasks": { _ in
                let success = try await provider.cancelAllTasks()
                return ["success": .bool(success)]
            },
            "getScheduledTasks": { _ in
                let tasks = try await provider.getScheduledTasks()
                return ["tasks": .array(tasks.map { .dictionary($0) })]
            },
            "completeTask": { payload in
                guard let taskId = payload["taskId"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: taskId")
                }
                let success = payload["success"]?.boolValue ?? true
                provider.completeTask(taskId: taskId, success: success)
                return [:]
            },
            "requestPermission": { _ in
                let granted = try await provider.requestPermission()
                return ["granted": .bool(granted)]
            },
        ]
    }
}

File: android/background-task/src/main/kotlin/io/rynbridge/backgroundtask/BackgroundTaskModule.kt
package io.rynbridge.backgroundtask

import io.rynbridge.core.*

class BackgroundTaskModule(provider: BackgroundTaskProvider) : BridgeModule {

    override val name = "backgroundTask"
    override val version = "0.1.0"
    override val actions: Map<String, ActionHandler> = mapOf(
        "scheduleTask" to { payload ->
            val taskId = payload["taskId"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: taskId")
            val type = payload["type"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: type")
            val interval = payload["interval"]?.intValue?.toInt()
            val delay = payload["delay"]?.intValue?.toInt()
            val requiresNetwork = payload["requiresNetwork"]?.boolValue ?: false
            val requiresCharging = payload["requiresCharging"]?.boolValue ?: false
            val success = provider.scheduleTask(taskId, type, interval, delay, requiresNetwork, requiresCharging)
            mapOf("taskId" to BridgeValue.string(taskId), "success" to BridgeValue.bool(success))
        },
        "cancelTask" to { payload ->
            val taskId = payload["taskId"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: taskId")
            val success = provider.cancelTask(taskId)
            mapOf("success" to BridgeValue.bool(success))
        },
        "cancelAllTasks" to { _ ->
            val success = provider.cancelAllTasks()
            mapOf("success" to BridgeValue.bool(success))
        },
        "getScheduledTasks" to { _ ->
            val tasks = provider.getScheduledTasks()
            mapOf("tasks" to BridgeValue.array(tasks.map { BridgeValue.dict(it) }))
        },
        "completeTask" to { payload ->
            val taskId = payload["taskId"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: taskId")
            val success = payload["success"]?.boolValue ?: true
            provider.completeTask(taskId, success)
            emptyMap()
        },
        "requestPermission" to { _ ->
            val granted = provider.requestPermission()
            mapOf("granted" to BridgeValue.bool(granted))
        }
    )
}
